src/history.coffee

```diff
@@ -24,60 +26,73 @@ class History
   groupChangesSinceCheckpoint: (checkpointId) ->
     checkpointIndex = @getCheckpointIndex(checkpointId)
     return false unless checkpointIndex?
+    hasSeenChanges = false
     for entry, i in @undoStack by -1
       break if i is checkpointIndex
       if @undoStack[i] instanceof Checkpoint
-        @undoStack.splice(i, 1)
+        @undoStack.splice(i, 1) if hasSeenChanges
```

Comment:
So this allows us to preserve checkpoints that follow changes. Is the preserved checkpoint always at the top of the stack? If so, would it be possible to just start the loop at the penultimate stack entry to preserve the checkpoint at the top?